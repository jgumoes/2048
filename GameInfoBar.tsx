import { Text, View, Modal, StyleSheet } from 'react-native';
import Grid from './grid';
import { RectButton, gestureHandlerRootHOC } from 'react-native-gesture-handler';
import ResetSquare from './assets/restart-square.svg'
import UndoIcon from './assets/undo.svg'
import React, { useState } from 'react';

const ResetBoardModal = ({onYes, onNo}: {onYes: () => void, onNo: () => void}) => {
  return(
    <View style={styles.resetBoardModal}>
      <Text style={styles.resetBoardModalText}>Are you sure you want to reset?</Text>
      <View>
        <RectButton onPress={onNo}>
        <Text style={styles.resetBoardModalText}>no</Text>
        </RectButton>
        <RectButton onPress={onYes}>
        <Text style={styles.resetBoardModalText}>yes</Text>
        </RectButton>
      </View>
    </View>
  )
}

const WrappedResetBoardModal = gestureHandlerRootHOC(ResetBoardModal)

export default function GameInfoBar({grid}: {grid: Grid}) {
  const [showResetBoardModal, setShowResetBoardModal] = useState(false)
  const onResetSquarePress = () => setShowResetBoardModal(true)

  const onNoCallback = () => {console.log('user pressed no'); setShowResetBoardModal(false);}
  const onYesCallback = ()=> {console.log('user pressed yes'); grid.reset(); setShowResetBoardModal(false);}
  return(
    <View style={styles.gameInfoBar}>
      <Modal
        animationType='slide'
        transparent={true}
        visible={showResetBoardModal}
        onRequestClose={() => setShowResetBoardModal(false)}
      >
        <WrappedResetBoardModal onNo={onNoCallback} onYes={onYesCallback} />
      </Modal>
      <View>
        <Text>Score: {grid.currentScore}</Text>
        <Text>Undo Button Count: {grid.undoCount}</Text>
      </View>
      <RectButton onPress={()=>{grid.undo()}} >
        {/* <Text>{grid.undoCount}</Text> */}
        <UndoIcon width={100} height={100} />
      </RectButton>
      <RectButton onPress={onResetSquarePress} >
        <ResetSquare width={100} height={100} />
      </RectButton>
    </View>
  )
}

const styles = StyleSheet.create({
  gameInfoBar: {
    flexDirection: 'row'
  },
  resetBoardModal: {
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: 'gray',
    opacity: 0.8,
    height: '100%',
  },
  resetBoardModalText: {
    fontSize: 30
  },
  backButtonContainer: {
    flexDirection: "row",
    alignItems: 'center',
    // borderColor: '#1C274C',
    borderColor: '#000000',
    borderWidth: 10,
    borderStyle: 'solid',
  }
})
