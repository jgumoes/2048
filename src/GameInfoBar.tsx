import { Text, View, Modal, StyleSheet, Pressable } from 'react-native';
import Grid from './grid';
import { ResetSquare, UndoIcon } from './SVGIcons'
import React, { useState } from 'react';
import { gridSides } from './GameBoard';
import { observer } from 'mobx-react-lite';

const ResetBoardModal = ({onYes, onNo}: {onYes: () => void, onNo: () => void}) => {
  return(
    <View style={styles.resetBoardModal}>
      <Text style={styles.resetBoardModalText}>Are you sure you want to reset?</Text>
      <View>
        <Pressable onPress={onNo}>
        <Text style={styles.resetBoardModalText}>no</Text>
        </Pressable>
        <Pressable onPress={onYes}>
        <Text style={styles.resetBoardModalText}>yes</Text>
        </Pressable>
      </View>
    </View>
  )
}

const GameInfoBar = observer(({grid}: {grid: Grid}) => {
  const [showResetBoardModal, setShowResetBoardModal] = useState(false)
  const onResetSquarePress = () => {
    if(!grid.isGameOver){setShowResetBoardModal(true)}
    else{grid.reset()}
  }

  const componentHeight = gridSides()/5
  console.log('gridSides', gridSides())
  const scoreFontSize = (30 / 475) * gridSides()

  const onNoCallback = () => {console.log('user pressed no'); setShowResetBoardModal(false);}
  const onYesCallback = ()=> {console.log('user pressed yes'); grid.reset(); setShowResetBoardModal(false);}
  return(
    <View style={styles.gameInfoBar}>
      <Modal
        animationType='slide'
        transparent={true}
        visible={showResetBoardModal}
        onRequestClose={() => setShowResetBoardModal(false)}
      >
        <ResetBoardModal onNo={onNoCallback} onYes={onYesCallback} />
      </Modal>
      <View style={styles.scoresContainer}>
        <Text style={[{fontSize: scoreFontSize}, styles.scoresText]} >Score: {grid.currentScore}</Text>
        <Text style={[{fontSize: scoreFontSize}, styles.scoresText]} >Undo Count: {grid.undoCount}</Text>
      </View>
      <View style={[styles.buttonsContainer]} >
        <Pressable
          testID='undoButton'
          onPress={()=>{grid.undo()}}
        >
          <UndoIcon width={componentHeight} height={componentHeight} />
        </Pressable>
        <Pressable
          testID='resetButton'
          onPress={onResetSquarePress}
        >
          <ResetSquare width={componentHeight} height={componentHeight} />
        </Pressable>
      </View>
    </View>
  )
})

export default GameInfoBar

const styles = StyleSheet.create({
  gameInfoBar: {
    flexDirection: 'row',
    justifyContent: 'space-between'
  },
  resetBoardModal: {
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: 'gray',
    opacity: 0.8,
    height: '100%',
  },
  resetBoardModalText: {
    fontSize: 30
  },
  scoresContainer: {
    justifyContent: 'space-around'
  },
  scoresText: {
    color: '#1C274C'
  },
  buttonsContainer: {
    flexDirection: 'row',
    justifyContent: 'flex-end',
  }
})
