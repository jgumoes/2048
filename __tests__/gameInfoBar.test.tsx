import { act, render, screen, userEvent } from '@testing-library/react-native'
import Grid, { Direction } from '../src/grid';
// import { OnlyPlace2Grid as Grid } from './helpers/gridMocks';
import importedTestGrids from './helpers/testGrids'
import GameInfoBar from '../src/GameInfoBar';

jest.mock('../src/SVGIcons', () => {
  return{
    ResetSquare: () => {
      return <view />
    },
    UndoIcon: () => {
      return <view />
    }
  }
})

describe('GameInfoBar', () => {
  let testGrid: Grid;
  beforeEach(()=>{
    testGrid = new Grid({grid: importedTestGrids.resultGrids.down[0]});
  })
  test('renders', () => {
    render(<GameInfoBar grid={testGrid} />);
    expect(screen.toJSON()).toMatchSnapshot()
    expect(screen.toJSON()).not.toBeNull()
  })

  test('renders undo button', () => {
    render(<GameInfoBar grid={testGrid} />)
    expect(screen.getByTestId('undoButton')).toBeOnTheScreen()
  })

  test('renders reset button', () => {
    render(<GameInfoBar grid={testGrid} />)
    expect(screen.getByTestId('resetButton')).toBeOnTheScreen()
  })
})

test('score displays and updates correctly', () => {
  const testScore = 40200
  const testGrid = new Grid({grid: importedTestGrids.numberedInputs[0], score: testScore})
  render(<GameInfoBar grid={testGrid} />)
  expect(screen.getByText(`Score: ${testScore}`)).toBeOnTheScreen()
  expect(screen.toJSON()).toMatchSnapshot();
  act(() => {testGrid.swipe(Direction.left)})
  const secondTestScore = testScore + importedTestGrids.scores.left[0]
  expect(screen.getByText(`Score: ${secondTestScore}`)).toBeOnTheScreen()
  expect(screen.toJSON()).toMatchSnapshot()
})

describe('undo button', () => {
  let testGrid: Grid;
  const startingUndoCount = 5
  const startingScore = 42069
  beforeEach(()=>{
    testGrid = new Grid({grid: importedTestGrids.numberedInputs[0], undoCount: startingUndoCount, score: startingScore})
  })
  describe('increments undo count', () => {
  
    test('when props are update', () => {
      render(<GameInfoBar grid={testGrid} /> )
      expect(screen.queryByText(`Undo Count: ${startingUndoCount}`)).toBeOnTheScreen()
      act(()=>{testGrid.swipe(Direction.left); testGrid.undo();})
      expect(screen.queryByText(`Undo Count: ${startingUndoCount + 1}`)).toBeOnTheScreen()
    })

    test('when button is pressed', async () => {
      render(<GameInfoBar grid={testGrid} /> )
      expect(screen.getByText(`Undo Count: ${startingUndoCount}`)).toBeOnTheScreen()
      const user = userEvent.setup()
      act(()=>{testGrid.swipe(Direction.left)})
      await user.press(screen.getByTestId("undoButton"))
      expect(screen.getByText(`Undo Count: ${startingUndoCount + 1}`)).toBeOnTheScreen()
    })
  })

  describe('returns score to previous value', () => {

    test('when props are updated', () => {
      render(<GameInfoBar grid={testGrid} /> )
      expect(screen.queryByText(`Score: ${startingScore}`)).toBeOnTheScreen()
      act(()=>{testGrid.swipe(Direction.left); testGrid.undo();})
      expect(screen.queryByText(`Score: ${startingScore}`)).toBeOnTheScreen()
    })

    test('when button is pressed', async () => {
      render(<GameInfoBar grid={testGrid} /> )
      expect(screen.queryByText(`Score: ${startingScore}`)).toBeOnTheScreen()
      const user = userEvent.setup()
      act(()=>{testGrid.swipe(Direction.left)})
      await user.press(screen.getByTestId("undoButton"))
      expect(screen.queryByText(`Score: ${startingScore}`)).toBeOnTheScreen()
    })
  })
})

test.todo('highscore changes depending on undo count') // highscore must be implemented first lol