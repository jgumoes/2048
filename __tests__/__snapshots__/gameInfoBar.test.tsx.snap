// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameInfoBar renders 1`] = `
<View
  style={
    {
      "flexDirection": "row",
      "justifyContent": "space-between",
    }
  }
>
  <Modal
    animationType="slide"
    hardwareAccelerated={false}
    onRequestClose={[Function]}
    transparent={true}
    visible={false}
  />
  <View
    style={
      {
        "justifyContent": "space-around",
      }
    }
  >
    <Text
      style={
        [
          {
            "fontSize": 45,
          },
          {
            "color": "#1C274C",
          },
        ]
      }
    >
      Score: 
      0
    </Text>
    <Text
      style={
        [
          {
            "fontSize": 45,
          },
          {
            "color": "#1C274C",
          },
        ]
      }
    >
      Undo Count: 
      0
    </Text>
  </View>
  <View
    style={
      [
        {
          "flexDirection": "row",
          "justifyContent": "flex-end",
        },
      ]
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      testID="undoButton"
    >
      <view />
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      testID="resetButton"
    >
      <view />
    </View>
  </View>
</View>
`;

exports[`score displays and updates correctly 1`] = `
<View
  style={
    {
      "flexDirection": "row",
      "justifyContent": "space-between",
    }
  }
>
  <Modal
    animationType="slide"
    hardwareAccelerated={false}
    onRequestClose={[Function]}
    transparent={true}
    visible={false}
  />
  <View
    style={
      {
        "justifyContent": "space-around",
      }
    }
  >
    <Text
      style={
        [
          {
            "fontSize": 45,
          },
          {
            "color": "#1C274C",
          },
        ]
      }
    >
      Score: 
      40200
    </Text>
    <Text
      style={
        [
          {
            "fontSize": 45,
          },
          {
            "color": "#1C274C",
          },
        ]
      }
    >
      Undo Count: 
      0
    </Text>
  </View>
  <View
    style={
      [
        {
          "flexDirection": "row",
          "justifyContent": "flex-end",
        },
      ]
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      testID="undoButton"
    >
      <view />
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      testID="resetButton"
    >
      <view />
    </View>
  </View>
</View>
`;

exports[`score displays and updates correctly 2`] = `
<View
  style={
    {
      "flexDirection": "row",
      "justifyContent": "space-between",
    }
  }
>
  <Modal
    animationType="slide"
    hardwareAccelerated={false}
    onRequestClose={[Function]}
    transparent={true}
    visible={false}
  />
  <View
    style={
      {
        "justifyContent": "space-around",
      }
    }
  >
    <Text
      style={
        [
          {
            "fontSize": 45,
          },
          {
            "color": "#1C274C",
          },
        ]
      }
    >
      Score: 
      40300
    </Text>
    <Text
      style={
        [
          {
            "fontSize": 45,
          },
          {
            "color": "#1C274C",
          },
        ]
      }
    >
      Undo Count: 
      0
    </Text>
  </View>
  <View
    style={
      [
        {
          "flexDirection": "row",
          "justifyContent": "flex-end",
        },
      ]
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      testID="undoButton"
    >
      <view />
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      testID="resetButton"
    >
      <view />
    </View>
  </View>
</View>
`;
